[JSC] Fix GetButterfly handling in DFG array allocation sinking

hyjorc1 opened the pull request:
#### 58dd8cd97ec1f23aafa4ea9a9866c718275fc48c
<pre>
[JSC] Fix GetButterfly handling in DFG array allocation sinking
<a href="https://bugs.webkit.org/show_bug.cgi?id=295191">https://bugs.webkit.org/show_bug.cgi?id=295191</a>
<a href="https://rdar.apple.com/154419571">rdar://154419571</a>

Reviewed by Keith Miller.

DFG&apos;s object allocation sinking previously mishandled GetButterfly nodes
when their base arrays were sunk and later materialized. Specifically, when
PutByVal or GetByVal triggered array escape, the corresponding GetButterfly
nodes were not correctly updated to depend on the materialized array, leading
to incorrect butterfly references and potential crashes.

This patch fixes the issue by:

1. Introducing a new Allocation::Kind::ArrayButterfly to model GetButterfly as
   a dependent allocation.
2. Adding ArrayButterflyPropertyPLoc to track butterfly-to-array relationships.
3. Implementing fixGetButterflyEscapees() to synchronize escape status between
   arrays and their butterflies.
4. Enforcing materialization ordering to ensure arrays materialize before their
   GetButterfly uses.
5. Updating GetButterfly lowering to depend on the correct materialized array.

This fix ensures that GetButterfly correctly accesses the butterfly of
materialized arrays.

Canonical link: <a href="https://commits.webkit.org/297146@main">https://commits.webkit.org/297146@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/967a2ad2c25ba30d73476863a9816510849b8f6c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30294 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20727 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116661 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60902 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30973 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38883 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84127 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113583 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24735 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99619 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64567 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24098 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17759 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60456 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103178 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94122 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17818 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119451 "Build is in progress. Recent messages:Printed configuration; Running apply-patch; Checked out pull request; Compiled WebKit") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109189 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37675 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27981 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93091 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95890 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92914 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23686 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37934 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15678 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33652 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37571 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43047 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/133464 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37233 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36060 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40572 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38941 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1030,14 +1030,17 @@ class ObjectAllocationSinkingPhase : public Phase {
             Node* base = node->child1().node();
```
Let's separate `GetBufferfly` and `GetArrayLength` handling, which makes code cleaner.
```
case GetButterfly {
   ...
   break;
}

case GetArrayLength: {
   ...
   break;
}
```

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2302,6 +2326,11 @@ class ObjectAllocationSinkingPhase : public Phase {
+                        // GetButterfly nodes will be removed during DCE since they become unused
+                        // after materialization replacement.
+                        break;
+
```
I think using `PhantomGetButterfly` would be simpler. Then, without a special handling of dependency, we can just keep it in the phantom tree of creation.

hyjorc1 replied:
This fix models GetButterfly as its own allocation (ArrayButterfly) to properly track its dependency on the base Array.

Constellation commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -2179,7 +2199,8 @@ class ObjectAllocationSinkingPhase : public Phase {
                     Node* escapee = m_materializationToEscapee.get(materialization);
-                    populateMaterialization(block, materialization, escapee);
+                    if (escapee->op() != GetButterfly)
+                        populateMaterialization(block, materialization, escapee);
```
If we have `PhantomGetButterfly`, you do not need to have this.

hyjorc1 replied:
This fix models GetButterfly as its own allocation (ArrayButterfly) to properly track its dependency on the base Array.

Constellation reviewed (commented):
We discussed offline with Yijia, and we think that we can make it simpler by introducing PhantomGetButterfly.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -296,6 +301,10 @@ class Allocation {
+        case Kind::ArrayButterfly:
+            out.print("ArrayButterfly"_s);
+            break;
+
```
Nit: I think you can just delete this whole function since EnumTraits will automatically stringily your enums now.

hyjorc1 replied:
Updated with `out.print(m_kind, "Allocation("_s);` since this function needs dump other non-enum properties.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1518,6 +1572,31 @@ class ObjectAllocationSinkingPhase : public Phase {
         }
 
+        auto fixGetButterflyEscapees = [&](auto& escapees) {
+            // Case (4): Remove GetButterfly if its base Array didn’t escape.
+            escapees.removeIf([&] (const auto& entry) {
+                return entry.key->op() == GetButterfly && !escapees.contains(entry.key->child1().node());
+            });
```
Why does this happen? What's an example of a node that would escape this? Shouldn't this just fall out if we've handled the relevant nodes in `handleNode`?

hyjorc1 replied:
This can happen in the `escapingOnEdge` case. 
```
for (BasicBlock* successorBlock : block->successors()) {
  if (!m_heapAtHead[successorBlock].isAllocation(entry.key)
      || m_heapAtHead[successorBlock].getAllocation(entry.key).isEscapedAllocation())
      mustEscape = true;
}
```
The test case `array-allocation-sink-upsilon-with-double-value.js` provides a graph:
```
BB#0
  Predecessors:
  Successors:
  Head
  Tail: Array

BB#2
  Predecessors: 0, 4
  Successors: 3, 4
  Head: Array
  Tail: Array, Butterfly

BB#4
  Predecessors: 2
  Successors: 2
  Head: Array, Butterfly
  Tail: Array, Butterfly
```
Here, at `BB#4`'s successorBlock `BB#2` doesn't have `Butterfly` at its `m_heapAtHead`. So, it `mustEscape`.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGObjectAllocationSinkingPhase.cpp:
```diff
@@ -1678,16 +1759,29 @@ class ObjectAllocationSinkingPhase : public Phase {
-                    reverseDependencies.add(field.value, NodeSet()).iterator->value.addVoid(entry.key);
-                    if (field.key.neededForMaterialization())
-                        myDependenciesForMaterialization.addVoid(field.value);
+        auto addDependency = [&](Node* a, Node* b, bool neededForMaterialization) {
+            dependencies.add(a, NodeSet()).iterator->value.addVoid(b);
+            reverseDependencies.add(b, NodeSet()).iterator->value.addVoid(a);
+            if (neededForMaterialization)
+                forMaterialization.add(a, NodeSet()).iterator->value.addVoid(b);
+        };
```
It seems like this would be easier to model (you wouldn't need a reverse dependency) if the ArrayButterfly `Allocation` held all the information about the properties/length of the `Array` and the `Array`'s `Allocation` only had one property, the Butterfly. Since that actually represents the runtime better I think more of this would just fall out.

We don't have to do that change here but can you file a `FIXME` to consider that in the future.

hyjorc1 replied:
Added a FIXME.

kmiller68 commented on Source/JavaScriptCore/dfg/DFGPromotedHeapLocation.cpp:
```diff
@@ -99,6 +99,10 @@ void printInternal(PrintStream& out, PromotedLocationKind kind)
+    case ArrayButterflyPropertyPLoc:
+        out.print("ArrayButterflyPropertyPLoc");
+        return;
+
```
Ditto. You should just be able to delete this whole function.

hyjorc1 replied:
Removed `void printInternal(PrintStream& out, PromotedLocationKind kind)`.

kmiller68 reviewed (approved):
r=me with some comments

hyjorc1 committed:
[JSC] Fix GetButterfly handling in DFG array allocation sinking
https://bugs.webkit.org/show_bug.cgi?id=295191
rdar://154419571

Reviewed by Keith Miller.

DFG's object allocation sinking previously mishandled GetButterfly nodes
when their base arrays were sunk and later materialized. Specifically, when
PutByVal or GetByVal triggered array escape, the corresponding GetButterfly
nodes were not correctly updated to depend on the materialized array, leading
to incorrect butterfly references and potential crashes.

This patch fixes the issue by:

1. Introducing a new Allocation::Kind::ArrayButterfly to model GetButterfly as
   a dependent allocation.
2. Adding ArrayButterflyPropertyPLoc to track butterfly-to-array relationships.
3. Implementing fixGetButterflyEscapees() to synchronize escape status between
   arrays and their butterflies.
4. Enforcing materialization ordering to ensure arrays materialize before their
   GetButterfly uses.
5. Updating GetButterfly lowering to depend on the correct materialized array.

This fix ensures that GetButterfly correctly accesses the butterfly of
materialized arrays.

Canonical link: https://commits.webkit.org/297146@main
